File: csharp-dotnet/src/Architecture.Shell.Cqrs/IPipelineBehavior.cs
namespace Architecture.Shell.Cqrs;

/// <summary>
/// Delegate representing the next step in the pipeline (behavior or handler).
/// </summary>
/// <typeparam name="TResponse">The response type returned by the pipeline.</typeparam>
/// <returns>A task that represents the asynchronous operation and returns the response.</returns>
public delegate Task<TResponse> RequestHandlerDelegate<TResponse>();

/// <summary>
/// Interceptor for cross-cutting concerns that wraps handler execution.
/// Behaviors execute in configured order to apply validation, authorization,
/// transactions, telemetry, caching, and other infrastructure concerns.
/// </summary>
/// <typeparam name="TRequest">The request type being processed.</typeparam>
/// <typeparam name="TResponse">The response type returned after processing.</typeparam>
/// <remarks>
/// <para><strong>Execution Pattern:</strong></para>
/// <code>
/// class ValidationBehavior&lt;TRequest, TResponse&gt; : IPipelineBehavior&lt;TRequest, TResponse&gt;
/// {
///     public async Task&lt;TResponse&gt; HandleAsync(TRequest request, RequestHandlerDelegate&lt;TResponse&gt; next, CancellationToken ct)
///     {
///         // Pre-handler logic
///         var validationResult = await _validator.ValidateAsync(request, ct);
///         if (validationResult.IsFailure)
///             throw new ValidationException(validationResult.Errors);
///
///         // Call next behavior or handler
///         var response = await next();
///
///         // Post-handler logic (if needed)
///         return response;
///     }
/// }
/// </code>
///
/// <para><strong>Common Behaviors (Recommended Order):</strong></para>
/// <list type="number">
/// <item>ValidationBehavior (order: 10) - Validate request payload</item>
/// <item>AuthorizationBehavior (order: 20) - Check permissions</item>
/// <item>UnitOfWorkBehavior (order: 30) - Open transaction for commands</item>
/// <item>TelemetryBehavior (order: 40) - Log request type, duration, status</item>
/// <item>CachingBehavior (order: 50) - Cache query results</item>
/// <item>ResilienceBehavior (order: 60) - Retry/timeout/circuit breaker</item>
/// </list>
///
/// <para><strong>Behavior Constraints:</strong></para>
/// <list type="bullet">
/// <item>Behaviors MUST call <c>next()</c> to continue pipeline unless explicitly short-circuiting</item>
/// <item>Behaviors MAY short-circuit by throwing exception or returning early</item>
/// <item>Behaviors execute in order determined by <see cref="Order"/> property</item>
/// <item>Behaviors can be conditionally applied using <see cref="IBehaviorMatcher"/></item>
/// </list>
/// </remarks>
public interface IPipelineBehavior<in TRequest, TResponse>
    where TRequest : IBaseRequest
{
    /// <summary>
    /// Gets the execution order for this behavior.
    /// Lower values execute first (outer behaviors).
    /// Higher values execute last (inner behaviors, closer to handler).
    /// </summary>
    int Order { get; }

    /// <summary>
    /// Handles the request by executing cross-cutting logic and calling the next behavior or handler.
    /// </summary>
    /// <param name="request">The request being processed.</param>
    /// <param name="continuation">
    /// Delegate to invoke the next behavior in the pipeline or the handler itself.
    /// MUST be called unless the behavior explicitly short-circuits the pipeline.
    /// </param>
    /// <param name="cancellationToken">Cancellation token for early termination.</param>
    /// <returns>The response after processing the request through the pipeline.</returns>
    Task<TResponse> HandleAsync(
        TRequest request,
        RequestHandlerDelegate<TResponse> continuation,
        CancellationToken cancellationToken);
}

File: typescript-nodejs/architecture-shell-cqrs/src/IPipelineBehavior.ts
import { BaseRequest } from './BaseRequest';

/**
 * RequestHandlerDelegate<TResponse> - Continuation delegate for pipeline chain
 *
 * Represents the next step in the pipeline (either another behavior or the handler itself).
 */
export type RequestHandlerDelegate<TResponse> = () => Promise<TResponse>;

/**
 * IPipelineBehavior<TRequest, TResponse> - Interceptor for cross-cutting concerns
 *
 * Wraps handler execution to apply cross-cutting concerns like validation,
 * authorization, transaction management, telemetry, and caching.
 *
 * @typeParam TRequest - The request type (must extend BaseRequest)
 * @typeParam TResponse - The response type
 *
 * @remarks
 * Behaviors execute in configured order (default: Validation → Authorization → Transaction → Telemetry → Resilience).
 * Behaviors MAY short-circuit pipeline by throwing exception or returning early.
 * Behaviors MUST call next() to continue pipeline unless explicitly short-circuiting.
 *
 * @example
 * ```typescript
 * export class ValidationBehavior<TRequest extends BaseRequest<TResponse>, TResponse>
 *   implements IPipelineBehavior<TRequest, TResponse> {
 *
 *   async handle(
 *     request: TRequest,
 *     next: RequestHandlerDelegate<TResponse>,
 *     signal: AbortSignal
 *   ): Promise<TResponse> {
 *     // Pre-handler logic
 *     const validationResult = await this.validator.validate(request);
 *     if (!validationResult.isValid) {
 *       throw new ValidationException(validationResult.errors);
 *     }
 *
 *     // Call next behavior or handler
 *     const response = await next();
 *
 *     // Post-handler logic (if needed)
 *     return response;
 *   }
 * }
 * ```
 *
 * @see {@link IBehaviorMatcher} - For filtering behaviors by request type
 */
export interface IPipelineBehavior<
  TRequest extends BaseRequest<TResponse>,
  TResponse
> {
  /**
   * Handles the request by wrapping the next behavior/handler in the pipeline
   *
   * @param request - The request being processed
   * @param next - Continuation delegate for the next step in pipeline
   * @param signal - Cancellation signal for early termination
   * @returns Promise resolving to the response
   * @throws May throw to short-circuit pipeline (e.g., validation/authorization failures)
   */
  handle(
    request: TRequest,
    next: RequestHandlerDelegate<TResponse>,
    signal: AbortSignal
  ): Promise<TResponse>;

  /**
   * Order determines behavior execution sequence
   * Lower numbers execute first (recommended: Validation=10, Authorization=20, Transaction=30, Telemetry=40, Caching=50)
   */
  order?: number;
}
